### ISSUE TITLE
10_1_Actor_Critic.ipynb 에 관한 질문입니다! (Question about 10_1 Actor_Critic)

### ISSUE THREAD

**Issue by wonchul-kim** [hunkim/ReinforcementZeroToAll#19]
네트워크 클래스에 create_op()에서 
policy_gain, value_loss, entropy 로 loss구성하셨는데.. 어떤 이론이 뒷받침 되는지

혹은 무엇을 reference로 하여 작성하셨는지 여주고 싶습니다!!
(특히 각각이 계산되는 식이 왜 저렇게 되는지 잘 이해가 가지 않습니다...)

I want to ask about s.t. in create_op() function of ActorCriticNetwork class.
There are values (policy_gain, value_loss, entropy) in create_op().
Could you explain how you get to those values or how those values are calculated?
Which reference did you use to get those values?

Thank you.

**Comment by kkweon** [comment]
## 요약
Actor Critic은 말그대로

1. Actor
2. Critic

두개가 존재합니다.

1. Actor은 state -> policy (= action probabilities)을 내놓는 함수이고
2. Critic은 state -> value를 내놓는 함수입니다.

이걸 보통 마지막 레이어를 추가해주는 식으로 구현이 되구요.
마지막 레이어를 제외한 네트워크는 같은 네트워크를 가지게 되구요.

이론은 actor-critic 이라는 이론입니다.

기본적으로 강화학습 접근 방식은

- Q-Learning과 같은 Value Iteration
- Policy Gradient의 Policy Iteration
 
두 가지가 존재합니다. 
이 두가지를 섞은 것이 Actor-Critic 알고리즘입니다.

계속 Policy를 수정해가면서
Value도 동시에 수렴하도록 한다고 생각하시면 될 것 같네요.
그래서 Value function이 생각한 값과 다르면 잘못된 Policy를 지양해주는 식으로 동작합니다.
그래서 Actor Critic이라는 말이 나오게 된 것이구요.

![image](https://cloud.githubusercontent.com/assets/2981167/26619241/6c44c370-4592-11e7-85eb-cdad2e20131e.png)


## Loss function
- Policy gain은 Policy Gradient를 통해서 아실 수 있습니다.
Advantage가 존재하며 이 Advantage에 따라 좋은 Action의 확률을 높이고 반대로 안좋은 Action은 하지 못하도록 합니다.
- Value loss 는 말그대로 Value function은 현재 state에 적절한 value를 output으로 내야 하기 때문에
Mean Squared Error로 실제 Value (= discounted rewards)와 비교를 하게 되는 것이지요.
- Entropy는 policy gradient는 on policy이기 때문에 exploration의 효과를 엔트로피를 통해서 줍니다. 즉 엔트로피가 높을 수록 exploration을 더 많이 하기 때문에 그렇게 맞춰 줍니다.

**Comment by wonchul-kim** [comment]
먼저, 친절한 답변 정말 감사합니다!!!

좀더 여쭈고 싶은 부분이 loss function입니다.
전체적인 loss function에서 각각 poilcy_gain, value_loss, entropy의 의의는 이해가 가는데
왜 저렇게 계산이 되는지는 궁금증이 있습니다.

1. 이전에 policy gradient에 대해서 공부를 했었고, loss function을 만들때  다음과 같이 해서
loglikelihood 를 취했습니다.
log_lik = -Y*tf.log(Y') - (1 - Y)*tf.log( 1 - Y') 해서 구성하였는데,
policy_gain을 보면, -Y*tf.log(Y) 만 되어 있는 것 같아서요,,, 뒷 부분은 필요가 없는건가요?


2. Entropy부분은... 잘 이해가 가지 않는데,,, 왜 exploration의 효과를 주나요?? (0.01을 곱해준 것은
hueristic한 수치인가요??)



공부하고 잇는데,,, 질문이 많아졌네요ㅜㅜ.ㅜ.ㅜ..

actor-critic을 하고나서 계속적으로 ddpg, a3c를 하려고 하는데,, 논문만을 보고는 코드를 짤수가 없는데,, 혹시
참고하신 자료같은거 추천해주실수 있으면 감사하겠습니다!!!


감사합니다!

2017년 5월 31일 오후 3:53, Mo Kweon <notifications@github.com>님이 작성:

**Comment by kkweon** [comment]
1. 적으신거는 binary cross entropy네요. policy gradient랑은 조금 다릅니다. policy gradient는 loss 가 없습니다. policy gain만 존재하고 maximize (= gradient ascent) 방법이 없기 때문에 마이너스로 minimize 문제로 바꾸는 것입니다.
보상의 기대값을 미분한 것이 바로 policy gain입니다. 그래서 보상의 기대값을 maximize하는 것이 objective이구요.

2. 0.01은 hyper parameter입니다. entropy는 엔트로피 의 의미 그대로 입니다. 엔트로피가 높다는 의미는 경우의 수가 많다는 뜻이죠. 경우의 수가 많도록 항상 열어두기 위해 포함하는 것입니다.

좀더 간단히 설명하자면 entropy는 probability가 0.5일때 최대가 됩니다.
즉 랜덤액션에 가까울수록 엔트로피의 값은 높음을 의미합니다.
그래서 엔트로피를 높여주는 것입니다. 너무 엔트로피 값이 높으면 당연히 이걸 최우선으로 하기 때문에 학습이 안되겠지요 항상 랜덤액션만 하려고 하니까요. 그래서 0.01값으로 적절하게 정해줍니다.

a3c라고 불리우는 것은 actor critic 그 자체고 변한건 없습니다.
성능이 더 좋아지진 않고 여러 환경을 통해서 학습 속도만 빨라질뿐입니다.
ddpg도 아주 조금 바꼈을 뿐입니다. 기본부터 차근차근 구현해보시면 언젠가는 쉽게 쉽게 가능하실 겁니다.
사실 a3c는 멀티쓰레딩이나 병렬처리 코딩이 어려운 것이기 때문에 
파이썬 병렬프로그래밍과 관련한 자료를 습득하셔야 겠죠.

저는 구글 검색해서 나오는 모든 글들을 적어도 5페이지까지 있는 글들을 전부 읽어봅니다.
너무 하나에 의존하기보다는 최대한 많이 읽어보시는게 가장 좋은 길입니다.
구글 페이지랭크 알고리즘은 최고의 알고리즘입니다. 검색만 하면 되니 편해지죠.
너무 하나의 글에서 다 모든걸 얻으려고 하지 마시라는 애기밖에 드릴게 없네요.

**Comment by wonchul-kim** [comment]
답변 정말 감사합니다!
도움이 많이 되었습니다!

2017-06-01 7:48 GMT+09:00 Mo Kweon <notifications@github.com>:
